### PR TITLE
improve Multiselect "single" functionality

### PR DESCRIPTION
**Pull request by caravinci**
**Description of changes**

Fulfils the requirements from ticket https://github.com/icgc-argo/dac-ui/issues/93 for `single` selection autocomplete dropdown. 
Preview Storybook: https://fix-multiselect-integration--argo-ui-storybook.netlify.app/?path=/story/uikit-form-multiselect--basic

**Type of Change**

- [x] Bug
- [x] Styling
- [ ] New Feature

**Checklist before requesting review:**

- Design (select one):
  - [x] Matches design:
    - component sizes, spacing, and styles
    - font size, weight, colour
    - spelling has been double checked
  - [ ] No design provided, screenshot of changes included in PR
  - [ ] No changes to UI
- UI Kit (select one):
  - [ ] New Components with storybook stories created
  - [x] Modified Existing components and updates stories
  - [ ] No new UIKit components or changes
- [ ] Feature is minimally responsive
- [x] Manual testing of UI feature
- [ ] Add copyrights to new files
- [x] Connected ticket to PR

### CODE REVIEW

File: uikit/form/MultiSelect/index.tsx [hunk 1]
```diff
@@ -215,60 +226,77 @@ const MultiSelect: React.ComponentType<{
...
-    event.persist();
+  const [focusState, setFocusState] = React.useState(contextValue?.focused);
+  const [searchString, setSearchString] = React.useState('');
 
-    const newValue = single ? [child.props.value] : uniq([...value, child.props.value]);
+  const hasError = contextValue?.error || !!error;
+  const isDisabled = contextValue?.disabled || disabled;
 
-    event.target = {
+  const createCustomEvent = (event, value?) => ({
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:281]: First, this helper standardises the event sent to the handlers for this custom component, and removes repeated code.
Some handlers were not getting this, making integration difficult/inconsistent.

File: uikit/form/MultiSelect/index.tsx [hunk 2]
```diff
@@ -299,89 +327,61 @@ const MultiSelect: React.ComponentType<{
   const handleNewItemClick = (event) => {
     const newValue = single ? [searchString] : uniq([...value, searchString]);
 
-    event.target = {
-      id,
-      name,
-      tagName: 'MULTISELECT',
-      type: `select-${single ? 'one' : 'multiple'}`,
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:306]: This bit of code is repeated all over the place... that's what the helper mentioned above is for.

File: uikit/form/MultiSelect/index.tsx [hunk 3]
```diff
@@ -299,89 +327,61 @@ const MultiSelect: React.ComponentType<{
...
 
-          e.target = {
-            value: newValue,
-            name,
-          };
-
-          setSearchString('');
-          onChange(e);
+          setSearchString(single ? searchString : '');
+          onChange(createCustomEvent(event, newValue));
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:353]: this change (basically the same as in all other event handlers) solves the integration bugs described in the ticket.

File: uikit/form/MultiSelect/index.tsx [hunk 4]
```diff
@@ -192,19 +181,41 @@ function Highlight({ string, searchText }) {
   }
 }
 
-const MultiSelect: React.ComponentType<{
+type MultiSelectProps = {
+  /**
+   * Aria Label
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:186]: This whole thing (former lines 195 through 245) simply abstract the `props` into a type, so Storybook plays better.
Added some extra documentation and improved the defaults.

File: uikit/form/MultiSelect/index.tsx [hunk 5]
```diff
@@ -406,44 +406,47 @@ const MultiSelect: React.ComponentType<{
...
         {showPlaceHolder ? (
-          <PlaceHolder className={clsx({ disabled, error, focused: focusState })}>
-            {placeholder}
+          <PlaceHolder className={clsx({ disabled: isDisabled, hasError, focused: focusState })}>
+            {placeholder || single ? 'Select one' : 'Add one or more...'}
           </PlaceHolder>
-        ) : single ? (
-          <SingleValue className={clsx({ disabled, error, focused: focusState })}>
-            {get(head(selectedItems), 'displayName')}
-          </SingleValue>
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:424]: this component is unnecessary and creates the duplicate item shown in the ticket as a bug
image

Instead, I'm making the input field itself (i.e. where you type) display the value, if there's one available/selected.

File: uikit/form/MultiSelect/index.tsx [hunk 6]
```diff
@@ -149,15 +147,6 @@ const SectionTitle = styled('li')`
...
 
-const SingleValue = styled('span')`
-  ${({ theme }) => css(theme.typography.paragraph)};
-  margin-right: 0.3em;
-
-  &.disabled {
-    color: ${({ theme }) => theme.multiSelect.disabledTextColor};
-  }
-`;
-
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:160]: This component retired and moved to Costa Rica 🏖️

File: uikit/form/MultiSelect/index.tsx [hunk 7]
```diff
@@ -103,13 +100,14 @@ const InputBox = styled(StyledInputWrapper)`
...
-  ${({ theme }) => css(theme.typography.paragraph)};
+const Input = styled<'input', { autoComplete: string; single?: boolean }>('input')`
+  ${({ theme }) => css(theme.typography.default)};
+  background-color: transparent;
   border: none;
   display: block;
   flex-grow: 1;
-  background-color: transparent;
-  padding-left: 5px;
+  padding: 0 ${({ single }) => (single ? 10 : 5)}px;
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:109]: This change handles the difference in styling, when you compare the `single` multiselect.

File: uikit/form/MultiSelect/index.tsx [hunk 8]
```diff
@@ -103,13 +100,14 @@ const InputBox = styled(StyledInputWrapper)`
   padding: 2px;
 `;
 
-const Input = styled<'input', { autoComplete: string }>('input')`
-  ${({ theme }) => css(theme.typography.paragraph)};
+const Input = styled<'input', { autoComplete: string; single?: boolean }>('input')`
+  ${({ theme }) => css(theme.typography.default)};
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:104]: This ensures the input wrapper is able to define the right font size, based on the `size` given to the component, without being overridden.

File: uikit/form/MultiSelect/index.tsx [hunk 9]
```diff
@@ -94,7 +91,7 @@ const Gap = styled('div')`
 `;
 
 const InputBox = styled(StyledInputWrapper)`
-  ${({ theme }) => css(theme.typography.default)};
+  cursor: text;
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:94]: it's a text field, cursor shouldn't be a pointer.

File: uikit/form/MultiSelect/index.tsx [hunk 10]
```diff
@@ -215,60 +226,77 @@ const MultiSelect: React.ComponentType<{
...
-  }
 
-  const handleItemClick = (child) => (event) => {
-    event.persist();
+  const [focusState, setFocusState] = React.useState(contextValue?.focused);
+  const [searchString, setSearchString] = React.useState('');
 
-    const newValue = single ? [child.props.value] : uniq([...value, child.props.value]);
+  const hasError = contextValue?.error || !!error;
+  const isDisabled = contextValue?.disabled || disabled;
```
Comment (caravinci) [inline uikit/form/MultiSelect/index.tsx:279]: ehh... no mutations. thank you.

**Review by caravinci — COMMENTED** [review]
Looks worse than it actually is. Comments to assist combing through this mess 👍

**Review by mistryrn — APPROVED** [review]
LGTM 👍 Appreciated all the comments explaining everything too!

**Pull request merged**